File: Data Structures/Tree-based-DSAs/Binary-tree/binary-search-tree.py
""" binary search tree Operations in python: """

class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

# we only use inorder traversal in this case:
def inorder(root):
    if root is None:
        return
    inorder(root.left)
    print(str(root.data) + "->", end=" ")
    inorder(root.right)

#insert a node:

def insert(node, data):
    #return a new node if the tree is empty:
    if node is None:
        return Node(data)
    
    #Traverse to the right place and insert the node:
    if data < node.data:
        node.left = insert(node.left, data)
    else:
        node.right = insert(node.right, data)
    return node

#find the inorder successor:
def minValueNode(node):
    current = node

    #find the left most leaf:
    while (current.left is not None):
        current = current.left
    return current
#deleting a node:

def deleteNode(root, data):
    if root is None:
        return root

    #find the node to be deleted:

    if data < root.data:
        root.left = deleteNode(root.left, data)
    elif data > root.data:
        root.right = deleteNode(root.right, data)

    #here we have found the key to be deleted:
    else:
        #if the node is only with one child or no child:
        if root.left is None:
            temp = root.right
            root = None
            return temp
        elif root.right is None:
            temp = root.left
            root = None
            return temp
        #if the node has 2 children, place the inorder successor in the position of 
        #the node to be deleted:
        temp = minValueNode(root.right)
        root.key = temp.key

        #delete the inorder successor:
        root.right = deleteNode(root.right, temp.key)

    return root

if __name__ == "__main__":
 root = None
 root = insert(root, 8)
 root = insert(root, 3)
 root = insert(root, 1)
 root = insert(root, 6)
 root = insert(root, 7)
 root = insert (root, 10)
 root = insert(root, 14)
 root = insert(root, 4)

 print("Inorder traveral-> ", end = " ")
 inorder(root)

 print("\nDeleting 10: \n")
 root = deleteNode(root, 10)

 print("Inorder traveral-> ", end = " ")
 inorder(root)






    

File: Data Structures/Tree-based-DSAs/Binary-tree/binary-search-tree.cpp
#include "iostream"

using namespace std;

typedef struct node node_t;
/* node definition: */
typedef struct node{
    int key;
    node_t* left = nullptr, * right = nullptr;
}node_t;

/* create a new node: */

node_t* newNode(int key){
    node_t* temp = (node_t* )malloc(sizeof(node_t ));
    temp->key = key;
    temp->left = temp->right = nullptr;
    return temp;
}

/* inorder traversal: */
void inorder(node_t* root){
    /* check whether root is null*/
    if (!root){
        return;
    }

    /* traverse left sub-tree:*/
    inorder(root->left);

    /* traverse root sub-tree:*/
    cout << root->key <<" ->";
    /* traverse right sub-tree*/

    inorder(root->right);

}

/* insert a node:*/

node_t* insert(node_t * node, int key){
    /* return a new node if key is empty:*/
    if(!node){
        return newNode(key);
    }

    /* traverse to the right place and insert the node:*/
    if (key < node->key){
        node->left = insert(node->left, key);
    }
    else{
        node->right = insert(node->right, key);
    }
    return node;
}

/* find the inorder successor:*/

node_t* minValNode(node_t* node){
    node_t* current = node;

    //find the left-most tree:
    while (current && current->left != nullptr){
        current = current->left;
    }
    return current;
}


/* deleting a node: */

node_t* deleteNode(node_t * root, int key){
    /* return if key is empty:*/

    if (!root){
        return root;
    }

    /* find the node to be deleted:*/
    if (key < root->key){
        root->left = deleteNode(root->left, key);
    }
    else if (key > root->key){
        root->right = deleteNode(root->right, key);
    }
    /* we  have found the node at this point: */
    else{

        /* if node has only 1 child or no child:*/
        if (!root->left){
            node_t* temp = root->right;
            free(root);
            root = nullptr;
            return temp;
        }
        else if (!root->right){
            node_t* temp = root->left;
            free(root);
            root = nullptr;
            return temp;
        }
        /* if the node has 2 children: */
        node_t* temp = minValNode(root->right);

        /* place the inorder successor in the position of the node to be deleted*/
        root->key = temp->key;

        /* delete the inorder successor: */
        root->right = deleteNode(root->right, temp->key);

    }
    return root;
}

/* driver code: */

int main(){
    node_t* root = nullptr;
    root = insert(root, 8);
    root = insert(root, 3);
    root = insert(root, 1);
    root = insert(root, 6);
    root = insert(root, 7);
    root = insert(root, 10);
    root = insert(root, 14);
    root = insert(root, 4);

    cout << " Inorder traversal: \n";
    inorder(root);

    cout << " After deleting 10: \n";
    root = deleteNode(root, 10);

    cout << " Inorder traversal: \n";
    inorder(root);


}
